packages/devtools_app/test/debugger/program_explorer_test.dart

```diff
@@ -4,55 +4,166 @@
+
+      when(fakeServiceManager.connectedApp!.isProfileBuildNow)
+          .thenReturn(false);
+      when(fakeServiceManager.connectedApp!.isDartWebAppNow).thenReturn(false);
+
+      setGlobal(ServiceConnectionManager, fakeServiceManager);
+      setGlobal(IdeTheme, IdeTheme());
+      setGlobal(NotificationService, NotificationService());
+    });
+
+    Future<TestProgramExplorerController> initializeProgramExplorer(
+      WidgetTester tester,
+    ) async {
+      final programExplorerController = TestProgramExplorerController(
+        initializer: (controller) {
+          final libraryNode =
+              VMServiceObjectNode(controller, 'fooLib', testLib);
+          libraryNode.script = testScript;
+          libraryNode.location = ScriptLocation(testScript);
+          controller.rootObjectNodesOverride.add(libraryNode);
+        },
+      );
+      final explorer = ProgramExplorer(controller: programExplorerController);
+      programExplorerController.initialize();
+      await tester.pumpWidget(
+        wrap(
+          Builder(
+            builder: explorer.build,
+          ),
+        ),
+      );
+      expect(programExplorerController.initialized.value, true);
+      expect(programExplorerController.rootObjectNodes.value.numNodes, 1);
+      return programExplorerController;
+    }
+
+    testWidgets('correctly builds nodes', (WidgetTester tester) async {
+      final programExplorerController = await initializeProgramExplorer(tester);
+      final libNode = programExplorerController.rootObjectNodes.value.first;
+      final outline = (await libNode.outline)!;
+
+      // The outline should only contain a single Class node.
+      expect(outline.length, 1);
+      final clsNode = outline.first;
+      expect(clsNode.object, const TypeMatcher<Class>());
+      expect(clsNode.name, testClassRef.name);
+
+      // The class should contain a function and a field.
+      expect(clsNode.children.length, 2);
+      for (final child in clsNode.children) {
+        if (child.object is Func) {
+          expect(child.object, testFunction);
+        } else if (child.object is Field) {
+          expect(child.object, testField);
+        } else {
+          fail('Unexpected node type: ${child.object.runtimeType}');
+        }
+      }
+    });
+
+    testWidgets(
+      'selection',
+      (WidgetTester tester) async {
+        final programExplorerController =
+            await initializeProgramExplorer(tester);
+        final libNode = programExplorerController.rootObjectNodes.value.first;
+
+        // No node has been selected yet, so the outline should be empty.
+        expect(programExplorerController.outlineNodes.value.isEmpty, true);
+        expect(programExplorerController.scriptSelection, isNull);
+        expect(programExplorerController.outlineSelection.value, isNull);
+
+        // Select the library node and ensure the outline is populated.
+        await programExplorerController.selectNode(libNode);
```

Comment:
since these are widget tests, we should be selecting via `tester.tap`, not via the controller. Otherwise, this is just testing the controller and not the UI

Reply:
Ah good point. Done.